klippy/extras/bed_mesh.py

```diff
@@ -0,0 +1,596 @@
+    def set_mesh(self, mesh):
+        # Assign the current mesh.  If set to None, no transform
+        # is applied
+        self.z_mesh = mesh
+        self.splitter.set_mesh(mesh)
+        # cache the current position before a transform takes place
+        self.last_position[:] = self.toolhead.get_position()
+    def get_z_factor(self, z_pos):
+        if z_pos >= self.fade_end:
+            return 0.
+        elif z_pos >= self.fade_start:
+            return (self.fade_end - z_pos) / self.fade_dist
+        else:
+            return 1.
+    def get_position(self):
+        # Return last, non-transformed position
+        if self.z_mesh is None:
+            # No mesh calibrated, so send toolhead position
+            return self.toolhead.get_position()
+        else:
+            # return current position minus the current z-adjustment
+            x, y, z, e = self.toolhead.get_position()
+            z_adjust = self.get_z_factor(z) * self.z_mesh.get_z(x, y)
+            return [x, y, z - z_adjust, e]
+    def move(self, newpos, speed):
+        factor = self.get_z_factor(newpos[2])
+        if self.z_mesh is None or not factor:
+            # No mesh calibrated, or mesh leveling phased out.
+            self.toolhead.move(newpos, speed)
+        else:
+            self.splitter.build_move(self.last_position, newpos, factor)
+            while not self.splitter.traverse_complete:
+                split_move = self.splitter.split()
+                if split_move:
+                    self.toolhead.move(split_move, speed)
+                else:
+                    raise self.gcode.error(
+                        "Mesh Leveling: Error splitting move ")
+        self.last_position[:] = newpos
+    cmd_BED_MESH_OUTPUT_help = "Retrieve interpolated grid of probed z-points"
+    def cmd_BED_MESH_OUTPUT(self, params):
+        if self.z_mesh is None:
+            self.gcode.respond_info("Bed has not been probed")
+        else:
+            self.calibrate.print_probed_positions(self.gcode.respond_info)
+            self.z_mesh.print_mesh(self.gcode.respond, self.horizontal_move_z)
+    cmd_BED_MESH_CLEAR_help = "Clear the Mesh so no z-adjusment is made"
+    def cmd_BED_MESH_CLEAR(self, params):
+        self.set_mesh(None)
+
+
+class BedMeshCalibrate:
+    ALGOS = ['lagrange', 'bicubic']
+    def __init__(self, config, bedmesh):
+        self.printer = config.get_printer()
+        self.bedmesh = bedmesh
+        self.probed_z_table = None
+        self.build_map = False
+        self.probe_params = {}
+        points = self._generate_points(config)
+        self._init_probe_params(config, points)
+        self.probe_helper = probe.ProbePointsHelper(config, self, points)
+        # Automatic probe:z_virtual_endstop XY detection
+        detect_offset = config.getboolean('detect_z_offset', False)
+        self.z_offset = None
+        if not detect_offset and config.has_section('stepper_z'):
```

Comment:
One additional comment - I don't think mesh bed leveling needs to do z_offset detection - the z_offset passed to finalize() should be the only z_offset needed (the value passed to finalize() is the nominal distance between the nozzle and bed when the probe triggers).

I suspect this code is derived from the bed_tilt.py code - but that code is trying to do something a bit weird.  If bed_tilt is used in combination with a "probe:z_virtual_endstop" then there is a feedback between the calculated tilt and the XY position where Z homing is performed - the bed_tilt.py code does "back flips" to handle that so the user can make an appropriate update to their printer.cfg.  I don't think this applies to bed mesh (the user is not expected to update their config with the results of the mesh).

-Kevin